src/IO/PeekableReadBuffer.cpp

```diff
@@ -175,7 +175,7 @@ bool PeekableReadBuffer::nextImpl()
 
     if (checkpoint_at_end)
     {
-        checkpoint.emplace(working_buffer.begin());
+        checkpoint.emplace(position());
```

Comment:
Previous approach didn't work in situation when we put the checkpoint on the end of internal buffer, but still have some pending data in sub-buffer, i.e. after `makeContinuousMemoryFromCheckpointToPos()` call.

Reply:
Also I have to add some unit-tests.